src/coreclr/jit/optimizer.cpp

```diff
@@ -9589,6 +9590,16 @@ void Compiler::optMarkLoopRemoved(unsigned loopNum)
         JITDUMP("Removed loop " FMT_LP " has one or more live children\n", loopNum);
     }
 
+    for (BasicBlock* const auxBlock : Blocks())
+    {
+        if (auxBlock->bbNatLoopNum == loopNum)
+        {
+            JITDUMP("Resetting loop number for " FMT_BB " from " FMT_LP " to " FMT_LP ".\n", auxBlock->bbNum, loopNum,
+                    loop.lpParent);
+            auxBlock->bbNatLoopNum = loop.lpParent;
+        }
+    }
+
```

Comment:
Comments:
1. This should not be inside `#ifdef DEBUG`; it should always run.
2. I'd put it before marking the loop removed.
3. Use `loop.LoopBlocks()` to only iterate over the loop blocks, not all the function blocks.

Reply:
> This should not be inside #ifdef DEBUG; it should always run.

Ah, missed that.